File: src/test/java/uk/gov/justice/hmpps/probationteams/repository/LocalDeliveryUnit2RepositoryTest.java
package uk.gov.justice.hmpps.probationteams.repository;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.security.test.context.support.WithAnonymousUser;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.context.transaction.TestTransaction;
import org.springframework.transaction.annotation.Transactional;
import uk.gov.justice.hmpps.probationteams.model.LocalDeliveryUnit2;
import uk.gov.justice.hmpps.probationteams.model.ProbationTeam;

import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

@ActiveProfiles("test")
@ExtendWith(SpringExtension.class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.MOCK)
@Transactional
@WithAnonymousUser
public class LocalDeliveryUnit2RepositoryTest {

    @Autowired
    private LocalDeliveryUnit2Repository repository;

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @Test
    public void testPersistLocalDeliveryUnit() {
        final var ldu = LocalDeliveryUnit2.builder().probationAreaCode("ABC").localDeliveryUnitCode("ABC123X").functionalMailbox("pqr@stu.ltd.uk").build();

        repository.save(ldu);
        TestTransaction.flagForCommit();
        TestTransaction.end();

        TestTransaction.start();
        final var optionalOfLDU = repository.findByProbationAreaCodeAndLocalDeliveryUnitCode("ABC", "ABC123X");

        assertThat(optionalOfLDU).isPresent();
        final var persistentLdu = optionalOfLDU.get();
        assertThat(persistentLdu.getId()).isNotNull();

        // Business key equality
        assertThat(persistentLdu).isEqualTo(ldu);

        assertThat(ldu.getCreateUserId()).isEqualTo("anonymous");

        // Check the db...
        Long count = jdbcTemplate.queryForObject("select count(*) from LOCAL_DELIVERY_UNIT2 where LOCAL_DELIVERY_UNIT_ID = ?", Long.class, persistentLdu.getId());
        assertThat(count).isEqualTo(1);
    }

    @Test
    public void testPersistLduWithProbationTeams() {
        final var ldu = lduWithProbationTeams();

        repository.save(ldu);
        TestTransaction.flagForCommit();
        TestTransaction.end();

        TestTransaction.start();
        final var optionalOfLdu = repository.findByProbationAreaCodeAndLocalDeliveryUnitCode("ABC", "ABC123Y");
        assertThat(optionalOfLdu).isPresent();
        final var persistentLdu = optionalOfLdu.get();
        assertThat(persistentLdu.getId()).isNotNull();

        assertThat(persistentLdu.getProbationTeams()).isEqualTo(lduWithProbationTeams().getProbationTeams());
    }

    private LocalDeliveryUnit2 lduWithProbationTeams() {
        return LocalDeliveryUnit2
                .builder()
                .probationAreaCode("ABC")
                .localDeliveryUnitCode("ABC123Y")
                .probationTeams(
                        Map.of(
                                "T1", ProbationTeam.builder().functionalMailbox("t1@team.com").build(),
                                "T2", ProbationTeam.builder().functionalMailbox("t2@team.com").build()
                        ))
                .build();
    }
}

File: src/test/java/uk/gov/justice/hmpps/probationteams/repository/LocalDeliveryUnit2RepositoryTest.kt
package uk.gov.justice.hmpps.probationteams.repository

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.fail
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.security.test.context.support.WithAnonymousUser
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.test.context.transaction.TestTransaction
import org.springframework.transaction.annotation.Transactional
import uk.gov.justice.hmpps.probationteams.model.LocalDeliveryUnit2
import uk.gov.justice.hmpps.probationteams.model.ProbationTeam
import uk.gov.justice.hmpps.probationteams.utils.uniqueLduCode
import java.util.*

@ActiveProfiles("test")
@ExtendWith(SpringExtension::class)
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.MOCK)
@Transactional
@WithAnonymousUser

class LocalDeliveryUnit2RepositoryTest(
        @Autowired val repository: LocalDeliveryUnit2Repository,
        @Autowired val jdbcTemplate: JdbcTemplate
) {

    @Test
    fun `Persist Local Delivery Unit`() {
        val lduCode = uniqueLduCode()
        val ldu = LocalDeliveryUnit2(
                probationAreaCode = "ABC",
                localDeliveryUnitCode = lduCode,
                functionalMailbox = "pqr@stu.ltd.uk")

        repository.save(ldu)
        TestTransaction.flagForCommit()
        TestTransaction.end()

        TestTransaction.start()
        val optionalOfLDU = repository.findByProbationAreaCodeAndLocalDeliveryUnitCode("ABC", lduCode)

        assertThat(optionalOfLDU).isPresent
        val persistentLdu = optionalOfLDU.get()
        assertThat(persistentLdu.id).isNotNull()


        // Business key equality
        assertThat(persistentLdu).isEqualTo(ldu)

        assertThat(ldu.createUserId).isEqualTo("anonymous")

        // Check the db...
        val count = lduCount(persistentLdu.id)
        assertThat(count).isEqualTo(1)
    }

    @Test
    fun `Persist Local Delivery Unit with Probation Teams`() {
        val lduCode = uniqueLduCode()
        val ldu: LocalDeliveryUnit2 = lduWithProbationTeams(lduCode)

        repository.save(ldu)
        TestTransaction.flagForCommit()
        TestTransaction.end()

        TestTransaction.start()
        val optionalOfLdu = repository.findByProbationAreaCodeAndLocalDeliveryUnitCode("ABC", lduCode)
        assertThat(optionalOfLdu).isPresent
        val persistentLdu = optionalOfLdu.get()
        assertThat(persistentLdu.id).isNotNull()

        assertThat(persistentLdu.probationTeams).isEqualTo(lduWithProbationTeams(lduCode).probationTeams)

        val id = persistentLdu.id
        if (id != null) {
            assertThat(probationTeamCount(id)).isEqualTo(2)
        } else {
            fail("No ID for persisted LDU")
        }
    }


    @Test
    fun `Update Probation Team`() {
        val lduCode = uniqueLduCode()
        val ldu: LocalDeliveryUnit2 = lduWithProbationTeams(lduCode)
        repository.save(ldu)
        TestTransaction.flagForCommit()
        TestTransaction.end()

        TestTransaction.start()

        val persistentLduOpt = repository
                .findByProbationAreaCodeAndLocalDeliveryUnitCode("ABC", lduCode)

        if (persistentLduOpt.isEmpty) {
            fail("LDU not found")
        } else {
            val persistentLdu = persistentLduOpt.get()
            persistentLdu.probationTeams.remove("T1")
            persistentLdu.probationTeams["T2"] = ProbationTeam("zzz@zzz.com")
        }

        TestTransaction.flagForCommit()
        TestTransaction.end()

        TestTransaction.start()

        val updatedPersistentLduOpt = repository
                .findByProbationAreaCodeAndLocalDeliveryUnitCode("ABC", lduCode)

        if (updatedPersistentLduOpt.isEmpty) {
            fail("Updated LDU not found in db")
        } else {
            val updatedLdu = updatedPersistentLduOpt.get()
            assertThat(updatedLdu.probationTeams).isEqualTo(mutableMapOf("T2" to ProbationTeam("zzz@zzz.com")))

            val lduId = updatedLdu.id
            if (lduId != null) {
                assertThat(probationTeamCount(lduId)).isEqualTo(1)
            } else {
                fail("No ID for persisted LDU")
            }
        }
    }

    @Test
    fun `Deleting an LDU should delete its probation teams`() {
        val lduCode = uniqueLduCode()
        val ldu: LocalDeliveryUnit2 = lduWithProbationTeams(lduCode)

        val persistentLdu = repository.save(ldu)
        TestTransaction.flagForCommit()
        TestTransaction.end()

        // Check the db...
        val lduId = persistentLdu.id

        if (lduId != null) {
            TestTransaction.start()
            assertThat(lduCount(lduId)).isEqualTo(1)
            assertThat(probationTeamCount(lduId)).isEqualTo(2)
            repository.deleteById(lduId)
            TestTransaction.flagForCommit()
            TestTransaction.end()

            TestTransaction.start()

            assertThat(lduCount(lduId)).isEqualTo(0)
            assertThat(probationTeamCount(lduId)).isEqualTo(0)
        } else {
            Assertions.fail("The saved LDU did not have an Id")
        }
    }

    private fun probationTeamCount(lduId: UUID) =
            jdbcTemplate.queryForObject("""
                select count(*) 
                  from PROBATION_TEAM 
                 where LOCAL_DELIVERY_UNIT_ID = ?
                 """.trimIndent(), Long::class.java, lduId)

    private fun lduCount(lduId: UUID?) =
            jdbcTemplate.queryForObject("""
                select count(*) 
                  from LOCAL_DELIVERY_UNIT2 
                 where LOCAL_DELIVERY_UNIT_ID = ?
                 """.trimIndent(), Long::class.java, lduId)

    companion object {
        fun lduWithProbationTeams(lduCode: String): LocalDeliveryUnit2 = LocalDeliveryUnit2(
                probationAreaCode = "ABC",
                localDeliveryUnitCode = lduCode,
                probationTeams = mutableMapOf(
                        "T1" to ProbationTeam("t1@team.com"),
                        "T2" to ProbationTeam("t2@team.com")
                ))
    }
}